src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensRangeHandler.cs

```diff
@@ -64,5 +74,40 @@ public SemanticTokensRangeHandler(IGlobalOptionService globalOptions)
 
             return new RoslynSemanticTokens { Data = tokensData, IsFinalized = isFinalized };
         }
+
+        private async Task<bool> IsDataFinalizedAsync(Project project, CancellationToken cancellationToken)
+        {
+            // If the project's compilation isn't yet available, kick off a task in the background to
+            // hopefully make it available faster since we'll need it later to compute tokens.
+            if (!_projectIdToCompilation.ContainsKey(project.Id))
+            {
+                var compilationTask = project.GetCompilationAsync(cancellationToken);
+                _projectIdToCompilation.Add(project.Id, compilationTask);
+            }
+
+            // We use a combination of IsFullyLoaded + the completed project compilation as the metric
+            // for isFinalized. It may not be completely accurate but this is only a a temporary fix until
+            // workspace/semanticTokens/refresh is implemented.
+            var isFinalized = false;
+            var workspaceStatusService = project.Solution.Workspace.Services.GetRequiredService<IWorkspaceStatusService>();
+            var isFullyLoaded = await workspaceStatusService.IsFullyLoadedAsync(cancellationToken).ConfigureAwait(false);
+            if (isFullyLoaded)
```

Comment:
it looks like we call this service every time, which I don't think is necessarily an issue (if we're fully loaded it should return fairly fast).

However for implementation in regards to the above comment - it may be easier if the task you store in the dictionary is only completed when both isFullyLoaded = true and the getcompilation call after that has succeeded.  Then you can just move the `compilationTask.IsCompleted` up to the top of the method.

Reply:
That's a good optimization, thanks! I've updated the code.